Provide a native systemd unit file

vulpicastor opened the pull request:
- Add a systemd unit file, translated from the sysvinit script.
  (Trac: #962)
- Add Build-Depends on dh_systemd.

Signed-off-by: Lizhou Sha slz@mit.edu

andersk commented on debian/control:
```diff
@@ -2,7 +2,7 @@ Source: debathena-pyhesiodfs
 Priority: extra
 Maintainer: Debathena Project <debathena@mit.edu>
-Build-Depends: cdbs, debhelper, config-package-dev, python-support, python (>= 2.6~)
+Build-Depends: cdbs, debhelper, dh-systemd (>= 1.5), config-package-dev, python-support, python (>= 2.6~)
```
This is not satisfiable on precise, which does not have dh-systemd.

vulpicastor replied:
Since this patch is really intended for "future" (post-jessie and post-vivid) releases of Debian and Ubuntu, does this mean that we have to create a new `debian` branch for new releases only?

andersk replied:
Debathena builds simultaneously for all supported releases without branching, as this greatly simplifies the build infrastructure and release management process. One could imagine reasons to potentially make an exception, but this is not likely to be one of them, since there are easy workarounds, such as `dh-systemd (>= 1.5) | cdbs (<< 0.4.122~)`.

vulpicastor replied:
I see. That would allow `dh-systemd` Build-Depend to be dropped on Trusty and earlier.

vulpicastor commented on debian/debathena-pyhesiodfs.service:
```diff
@@ -0,0 +1,18 @@
+ExecStopPost=/bin/umount -l ${pyhesiodfs_dir}
+
+[Install]
+WantedBy=multi-user.target remote-fs.target openafs-client.service
```
I'm not sure if `WantedBy=openafs-client.service` is equivalent to the `# RequiredStart: openafs-client` in the LSB init script.

andersk replied:
No, `WantedBy` corresponds more to the `Default-Start` LSB header if anything ([documentation](http://www.freedesktop.org/software/systemd/man/systemd.unit.html#WantedBy= documentation)). I think it should be `multi-user.target` only.

vulpicastor replied:
I see. `openafs-client.service` do have `multi-user.target remote-fs.target` as its `WantedBy`, though. I originally only had `multi-user.target` (consistent with the sysvinit runlevels), and that did work fine.

vulpicastor commented on debian/debathena-pyhesiodfs.service:
```diff
@@ -0,0 +1,18 @@
+ExecStartPre=/bin/mkdir -p /mit
+ExecStartPre=/bin/chown root:pyhesiodfs ${pyhesiodfs_dir}
+ExecStartPre=/bin/chmod 770 ${pyhesiodfs_dir}
+ExecStart=/usr/bin/pyhesiodfs -f ${pyhesiodfs_dir} -o nonempty
```
Is checking for the running of FUSE in `ExecStartPre` necessary? I don't see an obvious way of doing it other than passing the relevant line in the LSB init script to `/bin/sh`.

andersk replied:
The `fuse` package provides `/lib/modules-load.d/fuse.conf` to load the `fuse` module, so we could just do `After=systemd-modules-load.service` or something.

vulpicastor replied:
We probably want both `Require` and `After`.

vulpicastor replied:
It does look that `multi-user.target` is `After=basic.target` is `After=sysinit.target` is `After=local-fs.target`, and `systemd-modules-load.service` is `Before=sysinit.target`, so we can get rid of the extra dependencies here.

andersk commented on debian/debathena-pyhesiodfs.service:
```diff
@@ -0,0 +1,18 @@
+Type=simple
+Environment="pyhesiodfs_dir=/mit"
+OOMScoreAdjust=-1000
+ExecStartPre=/bin/mkdir -p /mit
```
`${pyhesiodfs_dir}`?

vulpicastor replied:
Is defined on line 8. `${pyhesiodfs_dir}` for systemd unit file is equivalent to quote-enclosed `"$pyhesiodfs_dir"` for shell.

andersk replied:
What I’m pointing out is an inconsistency between the `mkdir` line, which uses `/mit` literally, and all the other lines that use the variable.

vulpicastor replied:
Oh, oops.

andersk commented on debian/debathena-pyhesiodfs.service:
```diff
@@ -0,0 +1,18 @@
+ExecStartPre=/bin/chown root:pyhesiodfs ${pyhesiodfs_dir}
+ExecStartPre=/bin/chmod 770 ${pyhesiodfs_dir}
+ExecStart=/usr/bin/pyhesiodfs -f ${pyhesiodfs_dir} -o nonempty
+ExecStopPost=/bin/umount -l ${pyhesiodfs_dir}
```
This `umount` fails because systemd has already stopped `pyhesiodfs` with `SIGTERM`. I think that’s sufficient and we should just get rid of this `ExecStopPost` line.

```
● debathena-pyhesiodfs.service - Hesiod automounter for Athena lockers
   Loaded: loaded (/lib/systemd/system/debathena-pyhesiodfs.service; enabled; vendor preset: enabled)
   Active: failed (Result: exit-code) since Wed 2015-05-13 00:20:10 EDT; 900ms ago
  Process: 29816 ExecStopPost=/bin/umount -l ${pyhesiodfs_dir} (code=exited, status=32)
 Main PID: 27418 (code=exited, status=0/SUCCESS)

May 13 00:19:44 fixed-disk systemd[1]: Starting Hesiod automounter for Athena lockers...
May 13 00:19:44 fixed-disk systemd[1]: Started Hesiod automounter for Athena lockers.
May 13 00:19:44 fixed-disk pyhesiodfs[27418]: Error: Invalid locker name: .Trash
May 13 00:19:44 fixed-disk pyhesiodfs[27418]: Error: Invalid locker name: .Trash-1000
May 13 00:20:10 fixed-disk systemd[1]: Stopping Hesiod automounter for Athena lockers...
May 13 00:20:10 fixed-disk umount[29816]: umount: /mit: not mounted
May 13 00:20:10 fixed-disk systemd[1]: debathena-pyhesiodfs.service: control process exited, code=exited status=32
May 13 00:20:10 fixed-disk systemd[1]: Stopped Hesiod automounter for Athena lockers.
May 13 00:20:10 fixed-disk systemd[1]: Unit debathena-pyhesiodfs.service entered failed state.
May 13 00:20:10 fixed-disk systemd[1]: debathena-pyhesiodfs.service failed.
```

vulpicastor replied:
Interesting, I must have overlooked this test case earlier. It does appear that systemd magically takes care of the unmount.

vulpicastor commented on debian/debathena-pyhesiodfs.service:
```diff
@@ -0,0 +1,16 @@
+[Unit]
+Description=Hesiod automounter for Athena lockers
+Requires=openafs-client.service
+After=local-fs.target network.target openafs-client.service
```
`openafs-client.service` has `Before=remote-fs.target umount.target` here. Do we need those? i.e. do we count pyhesiodfs as a "remote FS"?

vulpicastor committed:
Provide a native systemd unit file.

- Add a systemd unit file, translated from the sysvinit script.
  (Trac: #962)
- Add Build-Depends on dh_systemd.
- Add a systemd tmpfile to create /mit on demand.

vulpicastor commented on debian/debathena-pyhesiodfs.service:
```diff
@@ -0,0 +1,20 @@
+Description=Hesiod automounter for Athena lockers
+Requires=openafs-client.service
+After=local-fs.target network.target openafs-client.service
+Before=umount.target
```
`openafs-client.service` has `Before=remote-fs.target umount.target` here. Do we need those? i.e. do we count pyhesiodfs as a "remote FS"?

kaduk replied:
openafs-client needs to be Before unmount.target because openafs needs to do special handling to actually do the unmount cleanly and unload the kernel module; without that dependency systemd would just do 'umount /afs', the unit's shutdown entries would not get run, and we'd be stuck with the kernel module loaded.

with respect to remote-fs.target, there is an argument that we should be before it, since we do provide a filesystem namespace that other things might use.

vulpicastor commented on debian/debathena-pyhesiodfs.service:
```diff
@@ -0,0 +1,20 @@
+
+[Unit]
+Description=Hesiod automounter for Athena lockers
+Requires=openafs-client.service
```
Do we want the strong `Requires=` here, or would `Wants=` suffice? Do we want to enforce that if `afsd` was stopped, `pyhesiodfs` should also be stopped?

vulpicastor replied:
Also, do I `Wants=network-online.target` here, following `openafs-client.service`? `man systemd.special` seems to suggest that remote file systems should pull that in, and also use `After=network-online.target`.

vulpicastor commented on debian/debathena-pyhesiodfs.service:
```diff
@@ -0,0 +1,20 @@
+ExecStart=/usr/bin/pyhesiodfs -f ${pyhesiodfs_dir} -o nonempty
+
+[Install]
+WantedBy=multi-user.target
```
Are we also `WantedBy=remote-fs.target`, or should we just leave OpenAFS to deal with it?

kaduk replied:
I am inclined to leave out WantedBy=remote-fs.target, but the logic I use to justify it to myself doesn't really hold water (namely, that we are not a "real" remote filesystem since the data we serve is just symlinks, not actual file contents.  This doesn't really apply, since the contents of the symlinks are the result of (remote) DNS queries.  So I guess you should go ahead and make that change.

vulpicastor replied:
Actually, since I have `Requires=openafs-client.service`, `pyhesiodfs` will be killed when `openafs-client` stops for any reason. Can we do `WantedBy=openafs-client.service` instead?

andersk replied:
I think we should take out all declared relationships with `openafs-client`, on the theory that PyHesiodFS won’t actually fail with OpenAFS not running, and if you want non-broken symlinks, it’s your job to start both PyHesiodFS and OpenAFS.

`WantedBy=remote-fs.target` and even `Before=remote-fs.target` seems right, on the theory that the user might have an `@reboot` cron job that uses `/mit`. After looking at this further, it seems that Debian’s `cron.service` does not actually declare `After=remote-fs.target` like Fedora’s does (the latter turns out to be [scripts.mit.edu’s fault](https://bugzilla.redhat.com/show_bug.cgi?id=807029 scripts.mit.edu’s)). But the theory still seems logical…

vulpicastor replied:
In that case, I could change it to `Wants=openafs-client.service`. This is less stringent than `Requires` and would kickoff the starting of OpenAFS automatically if it hasn't already.

vulpicastor replied:
Actually never mind. I don't think we need any dependency on `openafs-client.service`, as @andersk pointed out.

vulpicastor commented:
OK, latest changes with systemd unit file pushed.

andersk commented:
+1

vulpicastor commented:
Closed by 8f47bf819ab9d43609f6dd505bffc1864710a95b
